proposed/2021/PackageNamespaces.md

```diff
@@ -0,0 +1,568 @@
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <!-- Add a namespace for Microsoft.* or NuGet.* under the nuget.org source. -->
+    <packageSource name="nuget.org">
+        <namespace id="Microsoft.*" />
+        <namespace id="NuGet.*" />
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+If a user would like to support namespaces on different package sources, they can repeat this process for each source.
+A user may also pin a specific package id instead of the complete namespace.
+Specific ids take precedence over the namespaces.
+
+**Example:**
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <!-- Add a namespace for Microsoft.* or NuGet.* under the nuget.org source. -->
+    <packageSource name="nuget.org">
+        <namespace id="Microsoft.*" />
+        <namespace id="NuGet.*" />
+    </packageSource>
+
+    <!-- Add a namespace for Contoso.* under the contoso source. -->
+    <packageSource name="contoso">
+        <namespace id="Contoso.*" />
+        <!-- Add a specific id to be download from  the contoso source. -->
+        <namespace id="Special.Package.With.An.Imperfect.Id" />
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+By default, namespaces will allow a package to match multiple namespaces and download with no precedence. In some cases, one might want to avoid ambiguous cases. That can be done by adding a `strict` flag.
+
+```xml
+<config>
+    <add key="namespaceMode" value="strict" />
+</config>
+```
+
+The strict flag provides a source pinning behavior by nature in which one namespace is allowed per package source. Entering a strict mode will provide users an error experience when there are namespace conflicts to which a user will be able to resolve by defining one namespace per package source configuration.
+
+An error for this experience might look like:
+
+```console
+NUXXXX: Package namespace {0} is listed on the following sources: {1}. Only one unique package namespace can be defined across sources.
+```
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+#### Technical background - how NuGet package installation works today
+
+This approach builds on top of the current package installation behavior as described in the [package installation process](https://docs.microsoft.com/nuget/concepts/package-installation-process).
+
+A few other important concepts:
+
+- global packages folder - The installation directory for all packages in PackageReference. Think of this as the `Program Files` equivalent. All packages are consumed from this location. A single package installation can be shared safely among many different builds. A project *does not own* a specific package installation.
```

Comment:
Will package namespaces require PackageReference, or will they work with `packages.config` as well?

Reply:
It will work for both. Primarily designed towards core workloads using `PackageReference`.